[query] refactor block matrix persistence to take backend context

catoverdrive opened the pull request:
I need to use the backend context while persisting in order to support this with user separation on the service. I also moved the methods onto the general backend in preparation for implementing this on the service backend.

patrick-schultz reviewed (changes requested):
A couple minor questions/suggestions, otherwise looks good

catoverdrive commented on hail/src/main/scala/is/hail/backend/Backend.scala:
```diff
@@ -26,6 +28,14 @@ abstract class Backend {
 
   def broadcast[T: ClassTag](value: T): BroadcastValue[T]
 
+  def persist(backendContext: BackendContext, id: String, value: BlockMatrix, storageLevel: String): Unit
+
+  def unpersist(backendContext: BackendContext, id: String): Unit
```
Yeah, I changed this with the idea that we'd just overload on the type of `value` for persist and that the ids would be unique across all types for `unpersist`

patrick-schultz replied:
Makes sense, thanks.

catoverdrive commented:
@patrick-schultz I'm sending this back to you because I made some pretty drastic changes trying to fix some errors.

The biggest non-refactoring change that the original this introduces is that we can't parse IR for a persisted block matrix reader if the persisted block matrix doesn't exist. (This makes some amount of sense if you consider that we also can't parse the IR for a native block matrix reader if the file doesn't exist.)

This led me down a rabbit hole of test failures since we're parsing IR/types a fair number of times, through the execution and after we get the result. After fiddling with it for a little bit, I removed UnpersistBlockMatrix. I'm not sure what I was thinking when I added it. I re-added an "unpersist" function to the backend to handle unpersisting BlockMatrices. It differs from the current Table/MatrixTable unpersist functions in that we only pass the id of the thing we want to unpersist, not the entire IR, since that's all we need.

patrick-schultz commented on hail/src/test/scala/is/hail/expr/ir/IRSuite.scala:
```diff
@@ -24,7 +24,7 @@ import is.hail.variant.{Call2, Locus}
 import org.apache.spark.sql.Row
 import org.json4s.jackson.{JsonMethods, Serialization}
-import org.testng.annotations.{DataProvider, Test}
+import org.testng.annotations.{BeforeClass, DataProvider, Test}
```
Doesn't look like this is being used.

patrick-schultz reviewed (changes requested):
Looks fine to me, just one unused import
